drv/lpc55-sprot-server/src/handler.rs

```diff
@@ -50,213 +54,125 @@ impl Handler {
             sprocket: crate::handler::sprockets::init(),
             update: Update::from(UPDATE_SERVER.get_task_id()),
             startup_state: StartupState {
-                supported: 1_u32 << (Protocol::V1 as u8 - 1),
+                supported: 1_u32 << (Protocol::V2 as u8 - 1),
                 bootrom_crc32: CRC32.checksum(&bootrom().data[..]),
-                buffer_size: BUF_SIZE as u32,
+                max_request_size: MAX_REQUEST_SIZE.try_into().unwrap_lite(),
+                max_response_size: MAX_RESPONSE_SIZE.try_into().unwrap_lite(),
             },
         }
     }
 
     /// Serialize and return a `SprotError::FlowError`
-    pub fn flow_error<'a>(&self, tx_buf: TxMsg<'a>) -> VerifiedTxMsg<'a> {
-        tx_buf.error_rsp(SprotError::FlowError)
+    pub fn flow_error(&self, tx_buf: &mut [u8]) -> usize {
+        let body = Err(SprotProtocolError::FlowError.into());
+        Response::pack(&body, tx_buf).unwrap_lite()
     }
 
-    pub fn handle<'a>(
+    pub fn handle(
         &mut self,
-        rx_buf: RxMsg,
-        mut tx_buf: TxMsg<'a>,
+        rx_buf: &[u8],
+        tx_buf: &mut [u8],
         stats: &mut RotIoStats,
-    ) -> Option<VerifiedTxMsg<'a>> {
-        // Parse the header and validate the CRC
-        let rx_msg = match rx_buf.parse() {
-            Ok(rxmsg) => rxmsg,
-            Err((header_bytes, msgerr)) => {
-                if msgerr == SprotError::NoMessage {
-                    // We were just returning a reply, so clocked out zeros
-                    // from the SP.
-                    return None;
-                }
-                ringbuf_entry!(Trace::ErrWithHeader(msgerr, header_bytes));
+    ) -> usize {
+        stats.rx_received = stats.rx_received.wrapping_add(1);
+        let request = match Request::unpack(rx_buf) {
+            Ok(request) => {
+                ringbuf_entry!(Trace::Req {
+                    protocol: rx_buf[0],
+                    body_type: rx_buf[1]
+                });
+                request
+            }
+            Err(e) => {
+                ringbuf_entry!(Trace::Err(e));
                 stats.rx_invalid = stats.rx_invalid.wrapping_add(1);
-                return Some(tx_buf.error_rsp(msgerr));
+                return Response::pack(&Err(e.into()), tx_buf).unwrap_lite();
```

Comment:
nit: it seems like this places some requirements on `tx_buf` to avoid a panic here. If I called with a `tx_buf` of length 0, for instance, it'd panic. It'd be nice to put this (and any other usage restrictions) in a comment on the function.

Alternatively, it looks like this is always called with a `&[u8; MAX_RESPONSE_SIZE]` -- if `tx_buf`'s type were `&[u8; MAX_RESPONSE_SIZE]` instead of `&[u8]` then I think you _couldn't_ misuse it in this way?

Reply:
> Alternatively, it looks like this is always called with a &[u8; MAX_RESPONSE_SIZE] -- if tx_buf's type were &[u8; MAX_RESPONSE_SIZE] instead of &[u8] then I think you couldn't misuse it in this way?

I think this is a good suggestion. However, this code is shared among request and response which have different max sizes. I can go ahead and use a const parameter on Msg to deal with that though, and will do so.

Reply:
Fixed by using an array. Good idea!